src/main/java/patching/Diff.java

```diff
@@ -175,52 +175,21 @@ public Diff getUndo() {
-                        newIntermediateDiffs.add(newDiff);
-
+                if (other.startIndex < current.startIndex) {
+                    // CASE 1a, 1b: Ins - Ins, Ins - Rmv
+                    if ((other.insertion && current.insertion) || (other.insertion && !current.insertion)) {
+                        transformType2(newIntermediateDiffs, current, other);
                     }
                     // CASE 1c: Rmv - Ins
-                    else if (!other.insertion && intermediate.insertion) {
-                        if ((other.startIndex + other.getLength()) > intermediate.startIndex) {
-                            int newStartLoc = intermediate.startIndex - (intermediate.startIndex - other.startIndex);
-                            Diff newDiff = new Diff(intermediate.insertion, newStartLoc, intermediate.changes);
-                            newIntermediateDiffs.add(newDiff);
-                        } else {
-                            int newStartLoc = intermediate.startIndex - other.getLength();
-                            Diff newDiff = new Diff(intermediate.insertion, newStartLoc, intermediate.changes);
-                            newIntermediateDiffs.add(newDiff);
-                        }
-
+                    else if (!other.insertion && current.insertion) {
+                        transformType3(newIntermediateDiffs, current, other);
                     }
                     // CASE 1d: Rmv - Rmv
-                    else if (!other.insertion && !intermediate.insertion) {
-                        if ((other.startIndex + other.getLength()) <= intermediate.startIndex) {
-                            int newStartLoc = intermediate.startIndex - other.getLength();
-                            Diff newDiff = new Diff(intermediate.insertion, newStartLoc, intermediate.changes);
-                            newIntermediateDiffs.add(newDiff);
-                        } else if ((other.startIndex + other.getLength()) >= (intermediate.startIndex
-                                + intermediate.getLength())) {
-                            // Do nothing
-                        } else {
-                            int overlap = other.startIndex + other.getLength() - intermediate.startIndex;
-                            int newStartLoc = intermediate.startIndex - other.getLength() + overlap;
-                            String newChanges = intermediate.changes.substring(overlap);
-                            Diff newDiff = new Diff(intermediate.insertion, newStartLoc, newChanges);
-                            newIntermediateDiffs.add(newDiff);
-                        }
+                    else if (!other.insertion && !current.insertion) {
+                        transformType4(newIntermediateDiffs, current, other);
                     }
                     // FAIL: Should never have been able to get here.
                     else {
```

Comment:
So it's actually there, but just got truncated. [link](https://github.com/CodeCollaborate/ClientCore/blob/patching-update/src/main/java/patching/Diff.java#L196)